python/pyspark/ml/functions.py

```diff
@@ -106,6 +152,597 @@ def array_to_vector(col: Column) -> Column:
+    # |[0.0, 0.0, 0.0, 0...|[-13.511008, 8.84...|
+    # |[0.0, 0.0, 0.0, 0...|[-5.3957458, -2.2...|
+    # |[0.0, 0.0, 0.0, 0...|[-7.2014456, -8.8...|
+    # |[0.0, 0.0, 0.0, 0...|[-19.466187, -13....|
+    # |[0.0, 0.0, 0.0, 0...|[-5.7757926, -7.8...|
+    # +--------------------+--------------------+
+    ```
+
+    Example (scalar column):
+
+    Input DataFrame has a single scalar column, which will be passed to the `predict` function as
+    a 1-D numpy array.
+    ```
+    import numpy as np
+    import pandas as pd
+    from pyspark.ml.functions import predict_batch_udf
+    from pyspark.sql.types import FloatType
+
+    df = spark.createDataFrame(pd.DataFrame(np.arange(100)))
+    df.show(5)
+    # +---+
+    # |  0|
+    # +---+
+    # |  0|
+    # |  1|
+    # |  2|
+    # |  3|
+    # |  4|
+    # +---+
+
+    def make_predict_fn():
+        def predict(inputs: np.ndarray) -> np.ndarray:
+            # inputs.shape = [batch_size]
+            # outputs.shape = [batch_size], return_type = FloatType()
+            return inputs * 2
+
+        return predict
+
+    times_two = predict_batch_udf(make_predict_fn,
+                                  return_type=FloatType(),
+                                  batch_size=10)
+
+    df = spark.createDataFrame(pd.DataFrame(np.arange(100)))
+    df.withColumn("x2", times_two("0")).show(5)
+    # +---+---+
+    # |  0| x2|
+    # +---+---+
+    # |  0|0.0|
+    # |  1|2.0|
+    # |  2|4.0|
+    # |  3|6.0|
+    # |  4|8.0|
+    # +---+---+
+    ```
+
+    Example (multiple scalar columns):
+
+    Input DataFrame has muliple columns of scalar values.  If the user-provided `predict` function
+    expects a single input, then the user must combine the multiple columns into a single tensor
+    using `pyspark.sql.functions.struct` or `pyspark.sql.functions.array`.
+    ```
+    import numpy as np
+    import pandas as pd
+    from pyspark.ml.functions import predict_batch_udf
+    from pyspark.sql.functions import struct
+
+    data = np.arange(0, 1000, dtype=np.float64).reshape(-1, 4)
+    pdf = pd.DataFrame(data, columns=['a','b','c','d'])
+    df = spark.createDataFrame(pdf)
+    # +----+----+----+----+
+    # |   a|   b|   c|   d|
+    # +----+----+----+----+
+    # | 0.0| 1.0| 2.0| 3.0|
+    # | 4.0| 5.0| 6.0| 7.0|
+    # | 8.0| 9.0|10.0|11.0|
+    # |12.0|13.0|14.0|15.0|
+    # |16.0|17.0|18.0|19.0|
+    # +----+----+----+----+
+
+    def make_predict_fn():
+        def predict(inputs: np.ndarray) -> np.ndarray:
+            # inputs.shape = [batch_size, 4]
+            # outputs.shape = [batch_size], return_type = FloatType()
+            return np.sum(inputs, axis=1)
+
+        return predict
+
+    sum_rows = predict_batch_udf(make_predict_fn,
+                                 return_type=FloatType(),
+                                 batch_size=10,
+                                 input_tensor_shapes=[[4]])
+
+    df.withColumn("sum", sum_rows(struct("a", "b", "c", "d"))).show(5)
```

Comment:
It might be error-prone to auto merge sub-columns from `struct`. Shall we only support `array` if merging values into a single array is what users want?

Reply:
I think it's ok to support both, especially since it's hard to detect the difference from inside the UDF (since it will just get a pandas dataframe).  That said, I've updated the docs/examples to list `array` in priority over `struct` for these cases.

Reply:
What if the nested columns are not all numerical? `array` would throw an error at plan analysis time.

Reply:
Good point.  I've removed all references to `struct` now.